subt/artf_reporter.py

```diff
@@ -0,0 +1,93 @@
+message, where only "score_changed" has value 1 if it was correctly localized artifact within 5 meters radius (3D).
+
+The implemented solution is using shared array of items [type, position, source, scored]. Type and 3D position
+is the same as for DARPA report. "source" is identification of robot which detected artifact (for example "A1300L").
+"scored" has states: True/False/None. This corresponds to base station answer "score_changed"=1/"score_changed"=0/
+"unknown=no answer".
+
+If ArtifactReporter receives message with scored=None then first verifies if identical type and position is already
+in the pool. If yes, then received message is ignored. There still can be another report already confirmed as
+scored=True so it goes through identical types and if position is within RADIUS then it is rejected (it would need
+4th type - detected but ignored). Note, that in the rules "SubT_Challenge_Cave_Rules.pdf" is nothing about proximity
+of placed artifacts, so there can be two backpacks side by side. RADIUS is only "probabilistic", that it is rather
+duplicity than two artifacts placed side by side.
+
+Artifacts received with scored=True/False replace only scored=None in the pool. Not matching values should assert.
+
+There is a race condition between reporting artifact and spreading this info to other robots. It has to be done
+immediately otherwise other robots may use this time to report independently new found artifact. The only solution
+would be teambase, but then it would be the single point of failure.
+"""
+from osgar.node import Node
+from subt.trace import Trace, distance3D
+
+
+class ArtifactReporter(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        bus.register("artf_cmd", "artf_all")
+        self.repeat_report_sec = config.get('repeat_report_sec')
+        self.artf_xyz = []
+        self.artf_xyz_accumulated = []  # items are [type, position, source, flag]
+                                        # position = [x, y, z], flag(scored) = True/False/None
+
+    def publish_artf(self, artf_xyz):
+        count = 0
+        for artf_type, pos, src, scored in artf_xyz:
+            if scored is None:
+                if count == 0:
+                    print(self.time, "DETECTED:")
+                count += 1
+                ix, iy, iz = pos
+                print(" ", artf_type, ix/1000.0, iy/1000.0, iz/1000.0)
+                s = '%s %.2f %.2f %.2f\n' % (artf_type, ix/1000.0, iy/1000.0, iz/1000.0)
+                self.publish('artf_cmd', bytes('artf ' + s, encoding='ascii'))
+        if count > 0:
+            print('report completed')
+
+    def on_sim_time_sec(self, data):
+        if self.repeat_report_sec is None or self.sim_time_sec % self.repeat_report_sec != 0:
+            return
+        if len(self.artf_xyz) == 0:
+            return
+        # publish all artifacts only on regular update (to avoid LoRA cross-talk)
```

Comment:
This comment belongs to the `artf_all` call and not to the `publish_artf` call, right?